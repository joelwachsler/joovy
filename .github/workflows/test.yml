name: Test
on:
  push:
    tags:
      - '!**'
jobs:
  test:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Run tests
        run: yarn test

  lint:
    runs-on: self-hosted
    steps:
      - name: Setup build tools
        run: |
          sudo apt-get update
          sudo apt-get install build-essential autoconf automake g++ libtool -y

      - name: Checkout
        uses: actions/checkout@v2

      - name: Setup node
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Setup yarn
        run: npm install -g yarn

      - name: Setup cache
        uses: actions/cache@v2
        id: yarn-cache
        with:
          path: '**/node_modules'
          key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}

      - name: Install dependencies
        if: steps.yarn-cache.outputs.cache-hit != 'true'
        run: yarn

      - name: Lint
        run: yarn lint
