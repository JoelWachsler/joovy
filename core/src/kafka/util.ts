import { Kafka } from 'kafkajs'
import { defer, from, mapTo, mergeMap, Observable, tap } from 'rxjs'
import initConfig from '../config'
import logger from '../logger'
import { KMessage } from './kafka'

let kafka: Kafka

/**
 * Internal implementation detail, do not use.
 */
export const createKafkaClient = () => {
  if (!kafka) {
    const kafkaConfig = initConfig().kafka

    kafka = new Kafka({
      clientId: kafkaConfig.clientId,
      brokers: kafkaConfig.brokers,
    })
  }

  return kafka
}

type CreateConsumerArgs = {
  topic: string
  groupId: string
}

export const createConsumer = <T extends KMessage = KMessage>(args: CreateConsumerArgs) => {
  const { groupId, topic } = args

  const create = defer(() => {
    logger.info(`Creating consumer with the following args: ${JSON.stringify(args)}`)
    const consumer = createKafkaClient().consumer({ groupId })

    return from(consumer.connect()).pipe(mapTo(consumer))
  })

  const connectToTopic = create.pipe(mergeMap(consumer => {
    logger.info(`Subscribing to topic: ${topic}...`)
    return from(consumer.subscribe({ topic, fromBeginning: true })).pipe(
      tap(() => logger.info(`Subscribed to topic: ${topic}!`)),
      mapTo(consumer),
    )
  }))

  return connectToTopic.pipe(mergeMap(consumer => {
    return new Observable<T>(subscribe => {
      consumer.run({
        async eachMessage({ message, topic }) {
          const msgAsString = message.value?.toString()
          if (!msgAsString) {
            logger.error(`Invalid message: ${message} for topic: ${topic}`)
            return
          }

          logger.info(`New incoming message: ${JSON.stringify(msgAsString)} on topic: ${topic}`)

          subscribe.next(JSON.parse(msgAsString))
        },
      })
    })
  }))
}