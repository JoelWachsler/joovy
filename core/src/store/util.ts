import { defer } from 'rxjs'
import JEvent from '../jevent/JEvent'

const messageStore = new Map<string, Map<string, unknown>>()

export const getOrCreateObjectStore = (id: string) => {
  let store = messageStore.get(id)
  if (!store) {
    store = new Map()
    messageStore.set(id, store)
  }

  return store
}

export const removeObjectStore = (event: JEvent) => {
  return defer(() => {
    const channelId = event.message.channelId
    messageStore.delete(channelId)

    return event.result({ store: { event: 'removed', id: channelId } })
  })
}
