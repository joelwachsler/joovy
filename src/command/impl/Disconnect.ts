import { Message } from 'discord.js'
import { Environment } from '../../connectionHandler'
import { Command, ArgParser } from '../command'

export class Disconnect implements Command {
  argument = ArgParser.create('disconnect')
  helpText = 'Disconnects the bot from the current channel.'

  constructor(private env: Environment) {}

  async handleMessage(_: Message) {
    this.env.disconnect.next(null)
  }
}
