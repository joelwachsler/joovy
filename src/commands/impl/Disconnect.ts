import { Observable } from 'rxjs'
import JEvent, { ResultEntry } from '../../jevent/JEvent'
import { removePlaylist } from '../../playlist/Playlist'
import ArgParser from '../ArgParser'
import Command from '../command'

export default class Disconnect implements Command {
  argument = ArgParser.create('disconnect')
  helpText = 'Disconnects the bot from the current channel.'

  handleMessage(event: JEvent): Observable<ResultEntry> {
    return removePlaylist(event)
  }
}
