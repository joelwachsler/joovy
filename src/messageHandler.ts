import { mergeMap, Observable, of } from 'rxjs'
import { handle } from './commands/command'
import JEvent from './JEvent'

export const handleMessage = (event$: Observable<JEvent>): Observable<JEvent> => {
  return event$.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return of(event.withResult({ ignored: `${message.content} was sent by a bot` }))
      } else if (!message.content.startsWith('/')) {
        return of(event.withResult({ ignored: `${message.content} does not start with a slash` }))
      } else {
        return handle(event)
      }
    }),
  )
}
