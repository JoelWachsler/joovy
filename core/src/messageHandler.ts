import { Client } from 'discord.js'
import { catchError, filter, mergeMap, Observable } from 'rxjs'
import { handle } from './commands/command'
import { errorHandler } from './errorHandler'
import JEvent, { fromMessageKey } from './jevent/JEvent'
import { EmptyResult, Result } from './jevent/Result'
import { createConsumer, createProducer, KMessage, Topics } from './kafka/kafka'
import { logResult } from './resultLogger'

export const sendMessageEvent = (event: Observable<JEvent>): Observable<Result> => {
  return createProducer().pipe(mergeMap(producer => {
    return event.pipe(mergeMap(event => {
      return producer.send({
        event,
        message: createMessageEvent(event),
        topic: Topics.NewMessage,
      })
    }))
  }))
}

export const handleMessageEvents = (client: Client) => {
  const messageHandling = createConsumer<MessageEvent>({ groupId: 'joovy-message-handler', topic: Topics.NewMessage }).pipe(
    mergeMap(msg => fromMessageKey(client, msg.meta.messageKey)),
    handleMessage,
  )

  logResult('handleMessageEvents', messageHandling)
}

export const handleMessage = (event: Observable<JEvent>): Observable<Result> => {
  return event.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return event.result({ ignored: `${message.content} was sent by a bot` })
      } else if (!message.content.startsWith('/')) {
        return event.result({ ignored: `${message.content} does not start with a slash` })
      } else {
        return handle(event).pipe(
          filter(r => !(r instanceof EmptyResult)),
          catchError(err => errorHandler(event, err)),
        )
      }
    }),
  )
}

const createMessageEvent = (event: JEvent): MessageEvent => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'message-event',
      version: '1',
    },
    content: event.message.content,
  }
}

type MessageEvent = KMessage<'message-event', '1'> & { content: string }
