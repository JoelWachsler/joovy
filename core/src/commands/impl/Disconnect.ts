import { Client } from 'discord.js'
import { concat, mergeMap, Observable, of } from 'rxjs'
import JEvent, { fromMessageKey } from '../../jevent/JEvent'
import { Result } from '../../jevent/Result'
import { KMessage, Topics } from '../../kafka/kafka'
import { createConsumer } from '../../kafka/util'
import { removePlaylist } from '../../playlist/Playlist'
import { logResult } from '../../resultLogger'
import { removeObjectStore } from '../../store/util'
import ArgParser from '../ArgParser'
import Command from '../command'

export default class Disconnect implements Command {
  argument = ArgParser.create('disconnect')
  helpText = 'Disconnects the bot from the current channel.'

  handleMessage(event: JEvent): Observable<Result> {
    return sendDisconnectMessage(event, `manual disconnect by: ${JSON.stringify(event.message.author)}`)
  }
}

export const handleDisconnects = (client: Client) => {
  const disconnectHandler = createConsumer<DisconnectMessage>({ groupId: 'joovy-disconnect-handler', topic: Topics.Disconnect })
    .pipe(
      mergeMap(message => {
        return fromMessageKey(client, message.meta.messageKey)
          .pipe(mergeMap(event => disconnectFromChannel(of({ event, reason: message.reason }))))
      }),
    )

  logResult('disconnect', disconnectHandler)
}

export const disconnectFromChannel = (disconnectEvent: Observable<{ event: JEvent, reason: string }>) => {
  return disconnectEvent.pipe(mergeMap(({ event, reason }) => {
    const disconnectReason = event.result({ disconnect: reason })
    const playlistRemoval = removePlaylist(event)
    const removeChannelStore = removeObjectStore(event)
    const sendByeMessage = event.sendMessage('Bye!')

    return concat(
      disconnectReason,
      playlistRemoval,
      removeChannelStore,
      sendByeMessage,
    )
  }))
}

type DisconnectMessage = KMessage<'disconnect', '1'> & { reason: string }

const createDisconnectMessage = (event: JEvent, reason: string): DisconnectMessage => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'disconnect',
      version: '1',
    },
    reason,
  }
}

export const sendDisconnectMessage = (event: JEvent, reason: string) => {
  return event.factory.kafkaProducer().pipe(mergeMap(producer => {
    return producer.send({
      topic: Topics.Disconnect,
      event,
      message: createDisconnectMessage(event, reason),
    })
  }))
}
