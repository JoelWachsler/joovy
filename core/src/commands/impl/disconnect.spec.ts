import { merge } from 'rxjs'
import { handleMessage } from '../../messageHandler'
import { createTestEvent, handle, hot, resultAndMessages } from '../../test/util'
import { disconnectFromChannel } from './Disconnect'

test('should emit disconnect event', () => {
  const disconnect = createTestEvent({ content: '/disconnect' })

  const messages = handle(hot('a', { a: disconnect }))
  expect(messages).toMatchSnapshot()
})

test('should disconnect, reconnect and disconnect again without any problem', () => {
  const play = createTestEvent({ content: '/play test' })
  const disconnect = createTestEvent({ content: '/disconnect' })

  const messages = handleMessage(hot(         'ab-a', { a: play, b: disconnect }))
  const messages2 = disconnectFromChannel(hot('--b-', { b: disconnect }))
  expect(resultAndMessages(merge(messages, messages2))).toMatchSnapshot()
})

test('should not crash if disconnecting when not connected', () => {
  const disconnect = createTestEvent({ content: '/disconnect' })

  const messages = resultAndMessages(disconnectFromChannel(hot('a', { a: disconnect })))
  expect(messages).toMatchSnapshot()
})
