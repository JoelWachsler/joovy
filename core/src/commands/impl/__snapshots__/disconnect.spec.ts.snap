// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should disconnect, reconnect and disconnect again without any problem 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/disconnect",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "kafka": Object {
          "message": Object {
            "meta": Object {
              "messageKey": Object {
                "channelId": "testChannelId",
                "messageId": "0",
              },
              "name": "disconnect",
              "version": "1",
            },
            "reason": "manual disconnect by: {\\"username\\":\\"testuser\\",\\"bot\\":false,\\"id\\":\\"testAuthorId\\"}",
          },
          "topic": "joovy-disconnect",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "disconnect": "test",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "disconnected",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "store": Object {
          "event": "removed",
          "id": "testChannelId",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Bye!\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "trackingEvent",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 3,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 8,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 8,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should emit disconnect event 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/disconnect",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "kafka": Object {
          "message": Object {
            "meta": Object {
              "messageKey": Object {
                "channelId": "testChannelId",
                "messageId": "0",
              },
              "name": "disconnect",
              "version": "1",
            },
            "reason": "manual disconnect by: {\\"username\\":\\"testuser\\",\\"bot\\":false,\\"id\\":\\"testAuthorId\\"}",
          },
          "topic": "joovy-disconnect",
        },
      },
    },
  },
]
`;

exports[`should not crash if disconnecting when not connected 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "disconnect": "test",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "store": Object {
          "event": "removed",
          "id": "testChannelId",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Bye!\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;
